#include "mainwindow.h"
#include "./ui_mainwindow.h"

#include <QNetworkAccessManager>
#include <QNetworkReply>
#include <QNetworkRequest>

#include <QJsonDocument>
#include <QJsonObject>
#include <QUrl>
#include <QMessageBox>
#include <QTimer>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
    , m_network(new QNetworkAccessManager(this))
{
    ui->setupUi(this);

    connect(m_network, &QNetworkAccessManager::finished,
            this, &MainWindow::onLoginReply);

    ui->labelServerStatus->setText("ì„œë²„ ìƒíƒœ: ì—°ê²° í™•ì¸ ì¤‘...");
    checkServerHealth();

    auto timer = new QTimer(this);
    connect(timer, &QTimer::timeout, this, &MainWindow::checkServerHealth);
    timer->start(5000); // 5ì´ˆë§ˆë‹¤ /health í˜¸ì¶œ
}

//ì„œë²„ ìƒíƒœ ì²´í¬
void MainWindow::checkServerHealth()
{
    QUrl url("http://3.36.131.145:8080/health");
    QNetworkRequest req(url);
    m_network->get(req);
}

// ë²„íŠ¼ í´ë¦­ ìŠ¬ë¡¯
void MainWindow::on_btnLogin_clicked()
{
    const QString id = ui->lineEditId->text().trimmed();
    const QString pw = ui->lineEditPw->text();

    QUrl url("http://3.36.131.145:8080/login");
    QNetworkRequest req(url);
    req.setHeader(QNetworkRequest::ContentTypeHeader,
                  "application/json");

    QJsonObject obj;
    obj["id"] = id;
    obj["pw"] = pw;

    QByteArray body = QJsonDocument(obj).toJson(QJsonDocument::Compact);
    m_network->post(req, body);
}

// ë¡œê·¸ì¸ ì„œë²„ ì‘ë‹µ
void MainWindow::onLoginReply(QNetworkReply *reply)
{
    const QString path = reply->url().path(); // "/login" or "/health"

    // 1) ë„¤íŠ¸ì›Œí¬ ì—ëŸ¬
    if (reply->error() != QNetworkReply::NoError) {
        ui->labelServerStatus->setText("ì„œë²„ ìƒíƒœ: ë„¤íŠ¸ì›Œí¬ ì˜¤ë¥˜");
        QMessageBox::warning(this,
                             "ë„¤íŠ¸ì›Œí¬ ì˜¤ë¥˜",
                             "ì„œë²„ ì‘ë‹µ ì‹¤íŒ¨:\n" + reply->errorString());
        reply->deleteLater();
        return;
    }

    // 2) JSON íŒŒì‹±
    const QByteArray respData = reply->readAll();
    reply->deleteLater();

    QJsonParseError parseErr;
    QJsonDocument doc = QJsonDocument::fromJson(respData, &parseErr);
    if (parseErr.error != QJsonParseError::NoError || !doc.isObject()) {
        if (path == "/health") {
            ui->labelServerStatus->setText("ì„œë²„ ìƒíƒœ: ì‘ë‹µ íŒŒì‹± ì‹¤íŒ¨");
            QMessageBox::warning(this,
                                 "ì„œë²„ ìƒíƒœ",
                                 "ì‘ë‹µ íŒŒì‹± ì‹¤íŒ¨:\n" + parseErr.errorString());
        } else if (path == "/login") {
            QMessageBox::warning(this,
                                 "ë„¤íŠ¸ì›Œí¬ ì˜¤ë¥˜",
                                 "ì‘ë‹µ íŒŒì‹± ì‹¤íŒ¨:\n" + parseErr.errorString());
        }
        return;
    }

    // 3) JSON ë‚´ìš© ì²˜ë¦¬
    QJsonObject obj = doc.object();
    const bool ok = obj.value("ok").toBool();

    //health
    if (path == "/health") {
        if (ok)
            ui->labelServerStatus->setText("ğŸŸ¢  Online");
        else
            ui->labelServerStatus->setText("ğŸ”´ Offline");
        return;
    }

    // 4) /login ì²˜ë¦¬
    if (!ok) {
        QMessageBox::information(
            this,
            "ë¡œê·¸ì¸ ì‹¤íŒ¨",
            "ì•„ì´ë”” í˜¹ì€ ë¹„ë°€ë²ˆí˜¸ë¥¼ í™•ì¸í•˜ì„¸ìš”."
            );
        return;
    }

    const QString name = obj.value("name").toString();
    QMessageBox::information(
        this,
        "ë¡œê·¸ì¸ ì„±ê³µ",
        name + "ë‹˜ í™˜ì˜í•©ë‹ˆë‹¤!"
        );
}

MainWindow::~MainWindow()
{
    delete ui;
}
